Profiling of conversion checking

jespercockx opened the pull request:
This adds some counters for the various different phases of conversion checking. At the time of writing, running Agda with `--profile=conversion` gives the following statistics on the standard library:
```
comparing at eta record type            5,297
comparing at pi type                   22,805
comparing equal terms                 454,818
comparing irrelevant terms                960
comparing levels                       77,437
comparing sizes                         1,642
comparing sorts                       133,801
comparing terms                       967,728
comparing terms by reduction          297,738
eta-expanding at unit type                417
eta-expanding equation at record type   4,486
injectivity successful                  1,457
meta shortcut                         203,147
meta shortcut successful              200,427
trying injectivity                      3,157
trying to solve meta                  325,815
```
and the following counts for the Cubical library:
```
comparing at eta record type             36,857
comparing at face type                   27,303
comparing at interval type                3,215
comparing at path type                   19,354
comparing at pi type                    166,584
comparing equal terms                 1,197,759
comparing irrelevant terms                2,645
comparing levels                        132,665
comparing sorts                         149,990
comparing terms                       2,577,746
comparing terms by reduction            681,727
eta-expanding at unit type                  575
eta-expanding equation at record type    30,786
first-order shortcut                     84,816
injectivity successful                      706
meta shortcut                           405,738
meta shortcut successful                370,504
trying injectivity                        6,772
trying to solve meta                    585,458
```
In particular, this might help with answering questions such as "how often are we *actually* relying on eta-equality during conversion checking? (Answer: about 0.5% of the total number of conversion checks, about 1% of the non-trivial ones).

This is the first time I'm doing anything with the profiling machinery and I must say it was really easy to use. But if anything could be done in a more idiomatic way I'd be glad to hear. Also, should I add anything about this to the documentation?

jespercockx committed:
[ profile ] Add detailed counts for different phases of conversion checking

andreasabel commented:
> But if anything could be done in a more idiomatic way I'd be glad to hear. 

Not sure about more idiomatic, but when I read through the list it seems that items that are connected are not necessarily adjacent, and this makes it harder to understand.  Maybe name the descriptors more hierarchically?
E.g.
```
comparing at eta record type            5,297
...
eta-expanding at unit type                417
eta-expanding equation at record type   4,486
```
It feels that these entries are connected, but it is not immediately clear because they are far apart in the printout and use different wording.  
I would guess that the last two entries listed here are parts of the first.  This would suggest something like:
```
comparison at eta record type (total): 5,297
...
comparison at eta record type: eta-expansion at unit type: 417
comparison at eta record type: eta-expansion at non-unit record type: 4,486
```
Similarly, the other entries should be structured hierarchical.

> Also, should I add anything about this to the documentation?

Document the new flag variant for `--profile` in the release notes and user manual.

andreasabel commented on src/full/Agda/TypeChecking/Conversion.hs:
```diff
@@ -319,24 +324,26 @@ compareTerm' cmp a m n =
-                  c <- getRecordConstructor r
-                  -- Record constructors are covariant (see test/succeed/CovariantConstructors).
-                  compareArgs (repeat $ polFromCmp cmp) [] (telePi_ tel __DUMMY_TYPE__) (Con c ConOSystem []) m' n'
+              isSingleton <- isSingletonRecordModuloRelevance r ps
```
`isSingletonRecordModuloRelevance` is an expensive operation that shouldn't be done needlessly.  So it should not be pulled to the toplevel when there is a possibility that it is not used subsequently (meta/meta case, e.g.).

Also, it seems that changing the logic leads to new unsolved metas (see CI).  E.g. if one side is a meta (but shortcut failed), and now you succeed immediately due to singleton record, you might miss the solution of the meta.

andreasabel replied:
(Also, a change in the constraint solver should be a separate PR, if at all.  But I guess the semantics change was not intended.)

jespercockx replied:
Yes, I also noticed this. I'll revert the change to the conversion algorithm.

jespercockx commented:
I updated the strings and fixed the error (push incoming), here is the output on the standard library now:
```
compare                                      967,910
compare at eta record                          5,546
compare at eta-record: both neutral              331
compare at eta-record: both neutral at unit       69
compare at eta-record: eta-expanding           5,129
compare at eta-record: meta                       86
compare at function type                      22,812
compare by reduction                         297,769
compare by reduction: injectivity              3,157
compare by reduction: injectivity successful   1,457
compare equal                                454,758
compare irrelevant                               960
compare levels                                77,440
compare meta                                 325,835
compare meta shortcut                        203,118
compare meta shortcut successful             200,384
compare sizes                                  1,642
compare sorts                                133,819
```

(edit: I'm not sure why the total numbers are different now..)
(edit 2: I reran with --ignore-interfaces, and the problem seems to have disappeared.)

andreasabel commented:
This time, the failed test workflow seems to be the usual race...
- https://github.com/agda/agda/issues/5619

andreasabel commented on src/full/Agda/TypeChecking/Conversion.hs:
```diff
@@ -319,24 +324,26 @@ compareTerm' cmp a m n =
+                     -- if we are dealing with a singleton record,
+                     -- we can succeed immediately
+                     let profUnitEta = whenProfile Profile.Conversion $ tick "compare at eta-record: both neutral at unit"
+                     ifM (isSingletonRecordModuloRelevance r ps) (profUnitEta >> return ()) $ do
```
Is `(>> return ())` necessary?  `tick` already has type `m ()`.

andreasabel reviewed (approved):
Nice work!

Approved with one nitpick.

jespercockx committed:
[ re #6199 ] Remove `>> return ()`
